### ISSUE TITLE
How are previous tabs computed?

### ISSUE THREAD

**Issue by srsudar** [fwextensions/QuicKey#55]
When I'm jumping back in tabs, I'm sometimes surprised by the order of the jump. Whatever it is doing is a little too fancy for me. It looks like maybe it tries to special case multiple tabs with the same URL. This code from `recent-tabs.js`, for example:

```js
<snip>
			if (newTab) {
					// we found the same URL in a new tab, so copy over
					// the relevant keys and store it in the hash
					// using the new tab's ID.  also delete the URL
					// from the hash in case there are duplicate tabs
					// pointing at the same URL.
				newTab = createRecent(newTab, oldTab);
				newTabsByID[newTab.id] = newTab;
				newTabIDs.push(newTab.id);
				delete freshTabsByURL[oldTab.url];
			} else {
<snip>
```

Can you give me an overview of what it's trying to do? Is it essentially trying to build a stack of last-visited URLs, and when choosing a URL it chooses that URL in the freshest tab, rather than purely the last used tab, maybe?

**Comment by fwextensions** [comment]
No, that snippet is from when the stored list of recent tabs is refreshed with the current tab data, generally after the browser relaunches.  Tab and window IDs aren't stable across launches, so the code looks for recent tabs by their URL and then updates the tab and window IDs based on the current data.

If you're navigating back through the recent tabs with the *Switch to previous tab* command, it will go one by one through the recent tabs list, as long as you press the command again within 750ms of the last press (while the toolbar icon is inverted).  If you wait longer than that, then the tab you just navigated to is put on the top of the stack.  [This gist](https://gist.github.com/fwextensions/f2a79df5530354d6ef3058d3671c427a) has a more detailed example.

This timeout also applies when switching tabs manually.  So if you're on tab 1 and then quickly press <kbd>ctrl</kbd><kbd>tab</kbd> a bunch of times to get to tab 5, pressing the *Switch to previous tab* command will jump back to tab 1 (as long as you didn't pause for more than 750ms on any of the in-between tabs).  The goal is to track tabs you're actually looking at, and not just passing through.  

The timeout when using the *Switch to previous tab* command is due to the limitation of not being able to detect the keyup event when letting go of the alt/cmd key.  Otherwise, I'd just make it work like the standard alt-tab behavior.  I have something like that working in [this branch](https://github.com/fwextensions/QuicKey/tree/feature/popup-window-ui), at the expense of keeping a popup window open while navigating.

**Comment by fwextensions** [comment]
If that's the behavior you're looking for (which is different than how alt-tab works), you may be able to achieve it by just not debouncing the `addTab()` function in `background.js`, or changing the time to 0, as a quick test.  That will immediately move the newly focused tab to the top of the stack.  The 750ms cooldown would still be in effect, so if you pressed *Switch to previous* twice to get to B, and then waited a second, you could again press *Switch to previous* twice to get back to D, since C is the tab previously focused before B, and D before C.

**Comment by srsudar** [comment]
If I'm understanding, that is still slightly different than what I have in mind. In the **jump list** version, your tab jump state stays around as long as you don't switch tabs through a different mechanism than the jump list. (I'm changing my notation here so it's easier to type):

```
A -> B -> C -> [D]
```

I'm at `D`. I **jump back** twice and get back to `B`:

```
A -> [B] -> C -> D
```

At this point I can either **jump back** (sending me to `A`) or **jump forward** (sending me to `C`). This is true for as long as I stay at `B`, even beyond the 750ms cooldown. In other words, once you've jumped back you can jump forward again. The jump list is invalidated after a non-jump list navigation. So if I then used QuicKey to search for and visit tab `E`, the jump list would then be:

```
A -> B -> [E]
```

At this point I can jump back again, but no longer forward, and I've lost `C` and `D`.

I have this working in the fork. Your jumping might be better, as the jump list requires you to maintain a bit of state in your head. It feels very natural now that I've grown used to it in vim, though.

**Comment by fwextensions** [comment]
Yeah, removing debouncing wouldn't produce that behavior, since the stack would get reordered as you navigated.

Have you thought about rearranging the tabs to match the jump list order?  Then you'd have a persistent visual reminder of the list, and you could just use ctrl-tab/ctrl-shift-tab to move back and forth in that list.

**Comment by srsudar** [comment]
Reordering is an interesting idea. That sounds a bit heavy handed for what I'm after, but my workflow is obviously already pretty bespoke. Requisite [xkcd](https://xkcd.com/1172/) link.

I'm happy to create a PR if you'd want to merge it, to clarify. Just seemed like maybe not a knob that you think needs twiddling.
